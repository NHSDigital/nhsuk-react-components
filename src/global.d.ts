declare module 'nhsuk-frontend/packages/components/header/header.js';
declare module 'nhsuk-frontend/packages/components/checkboxes/checkboxes.js';
declare module 'nhsuk-frontend/packages/components/character-count/character-count.js';
declare module 'nhsuk-frontend/packages/components/tabs/tabs.js';
