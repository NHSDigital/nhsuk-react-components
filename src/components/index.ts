export * from './content-presentation/index.js';
export * from './form-elements/index.js';
export * from './layout/index.js';
export * from './navigation/index.js';
export * from './typography/index.js';
export * from './utils/index.js';
