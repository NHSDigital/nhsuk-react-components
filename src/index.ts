export * from './components';
export * from './patterns';
