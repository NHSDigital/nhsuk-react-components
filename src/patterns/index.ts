export * from './nav-a-z/index.js';
export * from './review-date/index.js';
