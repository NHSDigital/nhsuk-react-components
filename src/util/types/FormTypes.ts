import { HTMLProps } from 'react';
import { ErrorMessageProps } from '@components/form-elements/error-message/ErrorMessage';
import { HintTextProps } from '@components/form-elements/hint-text/HintText';
import { LabelProps } from '@components/form-elements/label/Label';

export interface FormElementProps {
  label?: string;
  labelProps?: LabelProps;
  error?: string | boolean;
  errorProps?: ErrorMessageProps;
  hint?: string;
  hintProps?: HintTextProps;
  formGroupProps?: HTMLProps<HTMLDivElement>;
  disableErrorLine?: boolean;
  id?: string;
  name?: string;
}
