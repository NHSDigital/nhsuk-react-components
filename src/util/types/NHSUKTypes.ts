export type NHSUKSize = 's' | 'm' | 'l' | 'xl';

export type InputWidth = '2' | '3' | '4' | '5' | '10' | '20' | '30' | 2 | 3 | 4 | 5 | 10 | 20 | 30;

export type CareCardType = 'non-urgent' | 'urgent' | 'emergency';

export type CardType = 'feature' | 'primary' | 'secondary' | CareCardType;

export type ColWidth =
  | 'full'
  | 'three-quarters'
  | 'one-half'
  | 'two-thirds'
  | 'one-third'
  | 'one-quarter';

export type HTMLAttributesWithData<T> = React.HTMLAttributes<T> & {
  [key: `data-${string}`]: unknown;
};
