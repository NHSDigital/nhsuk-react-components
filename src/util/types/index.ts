export * from './FormTypes.js';
export * from './LinkTypes.js';
export * from './NHSUKTypes.js';
export * from './TypeGuards.js';
