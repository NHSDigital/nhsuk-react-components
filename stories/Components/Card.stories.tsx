/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import { Card } from '../../src';
import { Meta, StoryObj } from '@storybook/react';
import { ColWidth } from '../../src/util/types/NHSUKTypes';

const meta: Meta<typeof Card> = {
  title: 'Components/Card',
  component: Card,
};
export default meta;
type Story = StoryObj<typeof Card>;

export const Standard: Story = {
  render: () => (
    <Card>
      <Card.Content>
        <Card.Heading>If you need help now but it&apos;s not an emergency</Card.Heading>
        <Card.Description>
          Go to <a href="stories#">111.nhs.uk</a> or <a href="stories#">call 111</a>
        </Card.Description>
      </Card.Content>
    </Card>
  ),
};

export const ClickableCard: Story = {
  args: { clickable: true },
  render: (args) => (
    <Card clickable={args.clickable}>
      <Card.Content>
        <Card.Heading className="nhsuk-heading-m">
          <Card.Link href="#">Introduction to care and support</Card.Link>
        </Card.Heading>
        <Card.Description>
          A quick guide for people who have care and support needs and their carers
        </Card.Description>
      </Card.Content>
    </Card>
  ),
};

export const CardWithImage: Story = {
  args: { clickable: true },
  render: (args) => (
    <Card clickable={args.clickable}>
      <Card.Image
        src="https://assets.nhs.uk/prod/images/A_0218_exercise-main_FKW1X7.width-690.jpg"
        alt=""
      />
      <Card.Content>
        <Card.Heading className="nhsuk-heading-m">
          <Card.Link href="#">Exercise</Card.Link>
        </Card.Heading>
        <Card.Description>
          Programmes, workouts and tips to get you moving and improve your fitness and wellbeing
        </Card.Description>
      </Card.Content>
    </Card>
  ),
};

export const FeatureCard: Story = {
  args: { feature: true },
  render: (args) => (
    <Card feature={args.feature}>
      <Card.Content>
        <Card.Heading>Feature card heading</Card.Heading>
        <Card.Description>Feature card description</Card.Description>
      </Card.Content>
    </Card>
  ),
};

export const CardGroup: Story = {
  args: { width: 'one-half' },
  argTypes: {
    width: {
      control: 'radio',
      options: ['one-half', 'one-third', 'one-quarter'],
    },
  },
  render: (args) => (
    <Card.Group>
      <Card.GroupItem width={args.width as ColWidth}>
        <Card clickable>
          <Card.Content>
            <Card.Heading className="nhsuk-heading-m">
              <Card.Link href="#">Introduction to care and support</Card.Link>
            </Card.Heading>
            <Card.Description>
              A quick guide for people who have care and support needs and their carers
            </Card.Description>
          </Card.Content>
        </Card>
      </Card.GroupItem>
      <Card.GroupItem width={args.width as ColWidth}>
        <Card clickable>
          <Card.Content>
            <Card.Heading className="nhsuk-heading-m">
              <Card.Link href="#">Help from social services and charities</Card.Link>
            </Card.Heading>
            <Card.Description>
              Includes helplines, needs assessments, advocacy and reporting abuse
            </Card.Description>
          </Card.Content>
        </Card>
      </Card.GroupItem>
      <Card.GroupItem width={args.width as ColWidth}>
        <Card clickable>
          <Card.Content>
            <Card.Heading className="nhsuk-heading-m">
              <Card.Link href="#">Money, work and benefits</Card.Link>
            </Card.Heading>
            <Card.Description>
              How to pay for care and support, and where you can get help with costs
            </Card.Description>
          </Card.Content>
        </Card>
      </Card.GroupItem>
      <Card.GroupItem width={args.width as ColWidth}>
        <Card clickable>
          <Card.Content>
            <Card.Heading className="nhsuk-heading-m">
              <Card.Link href="#">Care after a hospital stay</Card.Link>
            </Card.Heading>
            <Card.Description>
              Includes hospital discharge and care and support afterwards
            </Card.Description>
          </Card.Content>
        </Card>
      </Card.GroupItem>
    </Card.Group>
  ),
};
