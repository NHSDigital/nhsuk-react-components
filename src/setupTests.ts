import { TextDecoder, TextEncoder } from 'util';
import '@testing-library/jest-dom';
import { outdent } from 'outdent';

/**
 * Polyfill TextEncoder/TextDecoder for ReactDOM
 *
 * @see {@link https://github.com/jsdom/jsdom/issues/1695}
 */
Object.defineProperties(globalThis, {
  TextEncoder: {
    value: TextEncoder,
    configurable: true,
    enumerable: false,
    writable: true,
  },
  TextDecoder: {
    value: TextDecoder,
    configurable: true,
    enumerable: false,
    writable: true,
  },
});

/**
 * Polyfill Element methods for NHS.UK frontend error summary
 *
 * @see {@link https://github.com/jsdom/jsdom/issues/1695}
 */
Object.defineProperties(Element.prototype, {
  scrollIntoView: { value: jest.fn() },
});

/**
 * Polyfill `matchMedia()` for NHS.UK frontend tabs
 */
Object.defineProperty(globalThis, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: true,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    addEventListener: jest.fn(),
    removeListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

beforeEach(() => {
  const stylesheet = document.createElement('style');

  stylesheet.innerHTML = outdent`
    :root {
      --nhsuk-breakpoint-mobile: 20rem;
      --nhsuk-breakpoint-tablet: 40.0625rem;
      --nhsuk-breakpoint-desktop: 48.0625rem;
      --nhsuk-breakpoint-large-desktop: 61.875rem;
    }
  `;

  // Add styles for NHS.UK frontend checks
  document.head.appendChild(stylesheet);

  // Flag NHS.UK frontend as supported
  document.body.classList.add('nhsuk-frontend-supported');
});
