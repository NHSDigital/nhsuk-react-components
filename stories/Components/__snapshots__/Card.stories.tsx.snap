// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Card CardGroup smoke-test 1`] = `
<div class="nhsuk-grid-row nhsuk-card-group">
  <div class="nhsuk-grid-column-one-half nhsuk-card-group__item">
    <div class="nhsuk-card nhsuk-card--clickable">
      <div class="nhsuk-card__content">
        <h2 class="nhsuk-card__heading nhsuk-heading-m">
          <a class="nhsuk-card__link"
             href="#"
          >
            Introduction to care and support
          </a>
        </h2>
        <p class="nhsuk-card__description">
          A quick guide for people who have care and support needs and their carers
        </p>
      </div>
    </div>
  </div>
  <div class="nhsuk-grid-column-one-half nhsuk-card-group__item">
    <div class="nhsuk-card nhsuk-card--clickable">
      <div class="nhsuk-card__content">
        <h2 class="nhsuk-card__heading nhsuk-heading-m">
          <a class="nhsuk-card__link"
             href="#"
          >
            Help from social services and charities
          </a>
        </h2>
        <p class="nhsuk-card__description">
          Includes helplines, needs assessments, advocacy and reporting abuse
        </p>
      </div>
    </div>
  </div>
  <div class="nhsuk-grid-column-one-half nhsuk-card-group__item">
    <div class="nhsuk-card nhsuk-card--clickable">
      <div class="nhsuk-card__content">
        <h2 class="nhsuk-card__heading nhsuk-heading-m">
          <a class="nhsuk-card__link"
             href="#"
          >
            Money, work and benefits
          </a>
        </h2>
        <p class="nhsuk-card__description">
          How to pay for care and support, and where you can get help with costs
        </p>
      </div>
    </div>
  </div>
  <div class="nhsuk-grid-column-one-half nhsuk-card-group__item">
    <div class="nhsuk-card nhsuk-card--clickable">
      <div class="nhsuk-card__content">
        <h2 class="nhsuk-card__heading nhsuk-heading-m">
          <a class="nhsuk-card__link"
             href="#"
          >
            Care after a hospital stay
          </a>
        </h2>
        <p class="nhsuk-card__description">
          Includes hospital discharge and care and support afterwards
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Components/Card CardWithImage smoke-test 1`] = `
<div class="nhsuk-card nhsuk-card--clickable">
  <img class="nhsuk-card__img"
       alt
       src="https://assets.nhs.uk/prod/images/A_0218_exercise-main_FKW1X7.width-690.jpg"
  >
  <div class="nhsuk-card__content">
    <h2 class="nhsuk-card__heading nhsuk-heading-m">
      <a class="nhsuk-card__link"
         href="#"
      >
        Exercise
      </a>
    </h2>
    <p class="nhsuk-card__description">
      Programmes, workouts and tips to get you moving and improve your fitness and wellbeing
    </p>
  </div>
</div>
`;

exports[`Components/Card ClickableCard smoke-test 1`] = `
<div class="nhsuk-card nhsuk-card--clickable">
  <div class="nhsuk-card__content">
    <h2 class="nhsuk-card__heading nhsuk-heading-m">
      <a class="nhsuk-card__link"
         href="#"
      >
        Introduction to care and support
      </a>
    </h2>
    <p class="nhsuk-card__description">
      A quick guide for people who have care and support needs and their carers
    </p>
  </div>
</div>
`;

exports[`Components/Card FeatureCard smoke-test 1`] = `
<div class="nhsuk-card nhsuk-card--feature">
  <div class="nhsuk-card__content nhsuk-card__content--feature">
    <h2 class="nhsuk-card__heading nhsuk-card__heading--feature">
      Feature card heading
    </h2>
    <p class="nhsuk-card__description">
      Feature card description
    </p>
  </div>
</div>
`;

exports[`Components/Card Standard smoke-test 1`] = `
<div class="nhsuk-card">
  <div class="nhsuk-card__content">
    <h2 class="nhsuk-card__heading">
      If you need help now but it's not an emergency
    </h2>
    <p class="nhsuk-card__description">
      Go to
      <a href="stories#">
        111.nhs.uk
      </a>
      or
      <a href="stories#">
        call 111
      </a>
    </p>
  </div>
</div>
`;
